File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/SyntaxFacts.kt
package dev.phytolizer.minsk.analysis.syntax

object SyntaxFacts {
    fun binaryOperatorPrecedence(kind: SyntaxKind): Int = when (kind) {
        SyntaxKind.StarToken, SyntaxKind.SlashToken -> 5
        SyntaxKind.PlusToken, SyntaxKind.MinusToken -> 4
        SyntaxKind.EqualsEqualsToken, SyntaxKind.BangEqualsToken -> 3
        SyntaxKind.AmpersandAmpersandToken -> 2
        SyntaxKind.PipePipeToken -> 1
        else -> 0
    }

    fun unaryOperatorPrecedence(kind: SyntaxKind): Int = when (kind) {
        SyntaxKind.PlusToken, SyntaxKind.MinusToken, SyntaxKind.BangToken -> 6
        else -> 0
    }

    fun keywordKind(text: String): SyntaxKind = when (text) {
        "true" -> SyntaxKind.TrueKeyword
        "false" -> SyntaxKind.FalseKeyword
        else -> SyntaxKind.IdentifierToken
    }
}

File: nim/src/minsk/codeAnalysis/syntax/syntaxFacts.nim
import syntaxKind

proc binaryOperatorPrecedence*(kind: SyntaxKind): int =
  case kind
  of SyntaxKind.StarToken, SyntaxKind.SlashToken:
    5
  of SyntaxKind.PlusToken, SyntaxKind.MinusToken:
    4
  of SyntaxKind.EqualsEqualsToken, SyntaxKind.BangEqualsToken:
    3
  of SyntaxKind.AmpersandAmpersandToken:
    2
  of SyntaxKind.PipePipeToken:
    1
  else:
    0

proc unaryOperatorPrecedence*(kind: SyntaxKind): int =
  case kind
  of SyntaxKind.PlusToken,
      SyntaxKind.MinusToken,
      SyntaxKind.BangToken:
    6
  else:
    0

proc keywordKind*(text: string): SyntaxKind =
  case text
  of "true":
    SyntaxKind.TrueKeyword
  of "false":
    SyntaxKind.FalseKeyword
  else:
    SyntaxKind.IdentifierToken
